andela/ah-backend-prime
# #164798305 Article custom search and filtering

#### What does this PR do?
- Adds a search functionality and custom filtering of articles based on parameters
#### Description of tasks to be completed 
- Add a custom filter class specifying filter parameters
- Add default filter backends in settings
- Add filter permission classes in views
- Add a method to check for permissions
#### How should this be manually tested?
- Fetch this branch by running the command git fetch origin `ft-custom-filter164798305` in the terminal.
- Check into the branch git checkout `ft-custom-filter164798305`
- Create and activate the virtual environment
- Install dependencies `pip install -r requirements.txt`
- Make migrations by running `python manage.py makemigrations`
- Run the command `python manage.py migrate` to run the migrations
- Run the application `python manage.py runserver`
- To filter by title, use `api/v1/articles/?title=title`
- To filter by author, use `api/v1/articles/?author=username_of_author`
- To filter by tags, use `api/v1/articles/?tagsList=tag`
- To search by keyword within the article use `api/v1/articles/?search=keyword`
#### What are the relevant pivotal tracker stories?
[#164798305](https://www.pivotaltracker.com/story/show/164798305)
#### Screenshots
- Filter by title
![image](https://user-images.githubusercontent.com/36891299/56149948-5cf8b300-5fb6-11e9-8cc8-1a9b93effca2.png)
- Filter by tag
![image](https://user-images.githubusercontent.com/36891299/56149975-68e47500-5fb6-11e9-92f2-e8f399977ef8.png)
- Filter by author
![image](https://user-images.githubusercontent.com/36891299/56149999-7c8fdb80-5fb6-11e9-9839-16d4f2d6b451.png)
- Filter by keyword
![image](https://user-images.githubusercontent.com/36891299/56150026-8dd8e800-5fb6-11e9-9fd4-2fb8b5edbdf9.png)

[Finishes #164798305]
